File: leetcode/march2024/30_subarrays_with_k_different_integers.cpp
// INTUITION
// Standard Sliding Window Problem

// CODE

class Solution {
    public:
    int solve(vector<int>& nums, int k) {
        int n = nums.size(), res = 0;
        int start = 0, end = 0;
        unordered_map<int, int> count;

        while(end < n) {
            count[nums[end]]++;
            while(count.size() > k) {
                count[nums[start]]--;
                if(count[nums[start]] == 0)
                    count.erase(nums[start]);
                start++;
            }
            res += (end - start + 1);
            end += 1;
        }

        return res;
    }

    int subarraysWithKDistinct(vector<int>& nums, int k) {
        return solve(nums, k) - solve(nums, k - 1);
    }
};

File: leetcode/march2024/python/30_subarrays_with_k_different_integers.py
# INTUITION
# Standard Sliding Window Problem

# CODE

class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        
        def solve(nums, k):
            n = len(nums)
            res = start = end = 0
            count = {}
            while end < n:
                count[nums[end]] = count.get(nums[end], 0) + 1
                
                while len(count) > k:
                    count[nums[start]] -= 1
                    if count[nums[start]] == 0:
                        del count[nums[start]]
                    start += 1
                
                res += end - start + 1
                end += 1
            return res

        return solve(nums, k) - solve(nums, k - 1)